### ISSUE TITLE
[🐛 Bug]: 4.8.0 Regression(?) sessions hang indefinitely

### ISSUE THREAD

**Issue by ShadowLNC** [SeleniumHQ/selenium#11596]
### What happened?

At seemingly random points, the Selenium 4.7.2 Python 3.11 client bindings will block indefinitely when using Selenium Grid 4.8.0. (We did manage to reproduce with the 4.8.0 client bindings and did not explore further with that version.)

### Infrastructure
- Linux VM as the host
- VM runs a Selenium Hub container
- VM runs 1 each of Chrome and Firefox node containers (Chrome allows 7 sessions)
- Tests execute as Docker containers
  - New container spawned for each test
  - Maximum 5 parallel tests
  - Tests connect to the Selenium Hub container

This is replicated across multiple VMs (each with their own Selenium Hub instance), and has been seen in at least 3-4 different VMs (although they are all supposed to be identical).

We have an auto-update script which refreshes our Selenium Grid containers nightly. The following issue started around Wednesday last week, and coincides with the release of 4.8.0 - we did not see this issue previously.

One particular instance for which I have acquired logs resulted in three containers all hanging at the same time (2023-01-29 22:24:04 UTC+10:30). This appears to be coincidental, as I see instances where a single test hangs. We are considering rolling back to 4.7 in the meantime, depending on how often we see this issue (currently estimating 1 in 100-200 instances).

### How can we reproduce the issue?

```shell
This appears to most commonly occur when we are navigating to a URL (`WebDriver.get()`), although I also believe I saw it during a screenshot command. It also seems to be most common immediately following the initialisation of a session.
```

### Relevant log output

```shell
# Timestamps in UTC for 2023-01-29 (selenium/hub:4)
11:53:33.263 INFO [LocalDistributor.newSession] - Session request received by the Distributor: 
 [Capabilities {browserName: chrome, goog:chromeOptions: {args: [--disable-notifications, --disable-browser-side-navi..., --enable-automation, --disable-dev-shm-usage, --disable-gpu, --disable-accelerated-2d-ca..., --window-size=1920x1080, --headless], extensions: [], prefs: {credentials_enable_service: false, download.default_directory: /tmp/REDACTED..., download.prompt_for_download: false, profile.default_content_settings.popups: 0, profile.password_manager_enabled: false}}, pageLoadStrategy: normal, proxy: {httpProxy: REDACTED...., noProxy: [localhost, *.REDACTED, *.REDACTED], proxyType: manual, sslProxy: REDACTED....}}]
11:54:04.250 INFO [LocalDistributor.newSession] - Session created by the Distributor. Id: ed972858e53402bb24056e6a4d61d1e5 
 Caps: Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 109.0.5414.74, chrome: {chromedriverVersion: 109.0.5414.74 (e7c5703604da..., userDataDir: /tmp/.com.google.Chrome.Ebk3lP}, goog:chromeOptions: {debuggerAddress: localhost:37503}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: LINUX, proxy: {httpProxy: REDACTED...., noProxy: [localhost, *.REDACTED, *.REDACTED], proxyType: MANUAL, sslProxy: REDACTED....}, se:bidiEnabled: false, se:cdp: ws://172.18.0.4:4444/sessio..., se:cdpVersion: 109.0.5414.74, se:vnc: ws://172.18.0.4:4444/sessio..., se:vncEnabled: true, se:vncLocalAddress: ws://172.18.0.4:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
11:54:04.253 INFO [LocalDistributor.newSession] - Session created by the Distributor. Id: c9c93c5c1c141f12c1589093bee8d813 
 Caps: Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 109.0.5414.74, chrome: {chromedriverVersion: 109.0.5414.74 (e7c5703604da..., userDataDir: /tmp/.com.google.Chrome.fPDpbO}, goog:chromeOptions: {debuggerAddress: localhost:33925}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: LINUX, proxy: {httpProxy: REDACTED...., noProxy: [localhost, *.REDACTED, *.REDACTED], proxyType: MANUAL, sslProxy: REDACTED....}, se:bidiEnabled: false, se:cdp: ws://172.18.0.4:4444/sessio..., se:cdpVersion: 109.0.5414.74, se:vnc: ws://172.18.0.4:4444/sessio..., se:vncEnabled: true, se:vncLocalAddress: ws://172.18.0.4:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
11:54:04.383 INFO [LocalDistributor.newSession] - Session created by the Distributor. Id: 768de59fa597fe481a427ed1b5c98371 
 Caps: Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 109.0.5414.74, chrome: {chromedriverVersion: 109.0.5414.74 (e7c5703604da..., userDataDir: /tmp/.com.google.Chrome.H5hMa9}, goog:chromeOptions: {debuggerAddress: localhost:34493}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: LINUX, proxy: {httpProxy: REDACTED...., noProxy: [localhost, *.REDACTED, *.REDACTED], proxyType: MANUAL, sslProxy: REDACTED....}, se:bidiEnabled: false, se:cdp: ws://172.18.0.4:4444/sessio..., se:cdpVersion: 109.0.5414.74, se:vnc: ws://172.18.0.4:4444/sessio..., se:vncEnabled: true, se:vncLocalAddress: ws://172.18.0.4:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
12:00:16.293 INFO [GridModel.release] - Releasing slot for session id c9c93c5c1c141f12c1589093bee8d813
12:00:16.293 INFO [LocalSessionMap.lambda$new$0] - Deleted session from local Session Map, Id: c9c93c5c1c141f12c1589093bee8d813
12:00:16.428 INFO [LocalSessionMap.lambda$new$0] - Deleted session from local Session Map, Id: ed972858e53402bb24056e6a4d61d1e5
12:00:16.428 INFO [GridModel.release] - Releasing slot for session id ed972858e53402bb24056e6a4d61d1e5
12:00:16.562 INFO [GridModel.release] - Releasing slot for session id 768de59fa597fe481a427ed1b5c98371
12:00:16.562 INFO [LocalSessionMap.lambda$new$0] - Deleted session from local Session Map, Id: 768de59fa597fe481a427ed1b5c98371
12:18:21.100 INFO [LocalDistributor.newSession] - Session request received by the Distributor: 
 [Capabilities {browserName: chrome, goog:chromeOptions: {args: [--disable-notifications, --disable-browser-side-navi..., --enable-automation, --disable-dev-shm-usage, --disable-gpu, --disable-accelerated-2d-ca..., --window-size=1920x1080, --headless], extensions: [], prefs: {credentials_enable_service: false, download.default_directory: /tmp/REDACTED..., download.prompt_for_download: false, profile.default_content_settings.popups: 0, profile.password_manager_enabled: false}}, pageLoadStrategy: normal, proxy: {httpProxy: REDACTED...., noProxy: [localhost, *.REDACTED, *.REDACTED], proxyType: manual, sslProxy: REDACTED....}}]

# Timestamps in UTC+10:30 for 2023-01-29 (selenium/node-chrome:4)
Starting ChromeDriver 109.0.5414.74 (e7c5703604daa9cc128ccf5a5d3e993513758913-refs/branch-heads/5414@{#1172}) on port 22729
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
[1674993207.439][SEVERE]: bind() failed: Cannot assign requested address (99)
ChromeDriver was started successfully.
Starting ChromeDriver 109.0.5414.74 (e7c5703604daa9cc128ccf5a5d3e993513758913-refs/branch-heads/5414@{#1172}) on port 7465
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
[1674993212.652][SEVERE]: bind() failed: Cannot assign requested address (99)
Starting ChromeDriver 109.0.5414.74 (e7c5703604daa9cc128ccf5a5d3e993513758913-refs/branch-heads/5414@{#1172}) on port 6186
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
[1674993213.282][SEVERE]: bind() failed: Cannot assign requested address (99)
ChromeDriver was started successfully.
22:24:04.238 INFO [LocalNode.newSession] - Session created by the Node. Id: ed972858e53402bb24056e6a4d61d1e5, Caps: Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 109.0.5414.74, chrome: {chromedriverVersion: 109.0.5414.74 (e7c5703604da..., userDataDir: /tmp/.com.google.Chrome.Ebk3lP}, goog:chromeOptions: {debuggerAddress: localhost:37503}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: LINUX, proxy: Proxy(manual, http=REDACTED..., se:cdp: http://localhost:37503, se:cdpVersion: 109.0.5414.74, se:vncEnabled: true, se:vncLocalAddress: ws://172.18.0.4:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
22:24:04.239 INFO [LocalNode.newSession] - Session created by the Node. Id: c9c93c5c1c141f12c1589093bee8d813, Caps: Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 109.0.5414.74, chrome: {chromedriverVersion: 109.0.5414.74 (e7c5703604da..., userDataDir: /tmp/.com.google.Chrome.fPDpbO}, goog:chromeOptions: {debuggerAddress: localhost:33925}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: LINUX, proxy: Proxy(manual, http=REDACTED..., se:cdp: http://localhost:33925, se:cdpVersion: 109.0.5414.74, se:vncEnabled: true, se:vncLocalAddress: ws://172.18.0.4:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
22:24:04.375 INFO [LocalNode.newSession] - Session created by the Node. Id: 768de59fa597fe481a427ed1b5c98371, Caps: Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 109.0.5414.74, chrome: {chromedriverVersion: 109.0.5414.74 (e7c5703604da..., userDataDir: /tmp/.com.google.Chrome.H5hMa9}, goog:chromeOptions: {debuggerAddress: localhost:34493}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: LINUX, proxy: Proxy(manual, http=REDACTED..., se:cdp: http://localhost:34493, se:cdpVersion: 109.0.5414.74, se:vncEnabled: true, se:vncLocalAddress: ws://172.18.0.4:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
22:30:16.153 INFO [LocalNode.stopTimedOutSession] - Session id c9c93c5c1c141f12c1589093bee8d813 timed out, stopping...
22:30:16.292 INFO [SessionSlot.stop] - Stopping session c9c93c5c1c141f12c1589093bee8d813
22:30:16.292 INFO [LocalNode.stopTimedOutSession] - Session id ed972858e53402bb24056e6a4d61d1e5 timed out, stopping...
22:30:16.426 INFO [SessionSlot.stop] - Stopping session ed972858e53402bb24056e6a4d61d1e5
22:30:16.429 INFO [LocalNode.stopTimedOutSession] - Session id 768de59fa597fe481a427ed1b5c98371 timed out, stopping...
22:30:16.561 INFO [SessionSlot.stop] - Stopping session 768de59fa597fe481a427ed1b5c98371
Starting ChromeDriver 109.0.5414.74 (e7c5703604daa9cc128ccf5a5d3e993513758913-refs/branch-heads/5414@{#1172}) on port 1784
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.
[1674994701.127][SEVERE]: bind() failed: Cannot assign requested address (99)
22:48:22.080 INFO [LocalNode.newSession] - Session created by the Node. Id: 73b15bf59a5019f8f05b04ca47e1e9f3, Caps: Capabilities {acceptInsecureCerts: false, browserName: chrome, browserVersion: 109.0.5414.74, chrome: {chromedriverVersion: 109.0.5414.74 (e7c5703604da..., userDataDir: /tmp/.com.google.Chrome.gFv1hA}, goog:chromeOptions: {debuggerAddress: localhost:43713}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: LINUX, proxy: Proxy(manual, http=REDACTED..., se:cdp: http://localhost:43713, se:cdpVersion: 109.0.5414.74, se:vncEnabled: true, se:vncLocalAddress: ws://172.18.0.4:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify, webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:virtualAuthenticators: true}
```

### Operating System

Docker on a Ubuntu/Debian host (Python client base image python:3.11-slim-bullseye)

### Selenium version

(Client bindings) 4.7.2 Python 3.11

### What are the browser(s) and version(s) where you see this issue?

Chrome 109.0.5414.74 (selenium/node-chrome:4)

### What are the browser driver(s) and version(s) where you see this issue?

ChromeDriver 109.0.5414.74 (e7c5703604daa9cc128ccf5a5d3e993513758913-refs/branch-heads/5414@{#1172}) (selenium/node-chrome:4)

### Are you using Selenium Grid?

4.8.0 (selenium/hub:4)

**Comment by ShadowLNC** [comment]
Similar to #11593 but that references 4.6.0 and so I believe this is a different issue.

**Comment by ShadowLNC** [comment]
Hi,

I did manage to reproduce what seems to be the same error using this example (unfortunately I cannot provide access to the full project at this time). It may be possible to reduce it even more.

<details>
<summary>repro.py</summary>

```python
from datetime import datetime
import selenium
from selenium import webdriver

def log(val: str) -> None:
    """Log with datestamp."""
    print(datetime.now(), val)


def main() -> None:
    """Repro."""
    # wd = SCDriverChrome().create()
    opts = webdriver.ChromeOptions()
    opts.add_argument("--headless")
    wd = webdriver.Remote(
        "http://host.docker.internal:4444/wd/hub", desired_capabilities=opts.to_capabilities()
    )

    log(f"Browser Name: {wd.capabilities.get('browserName')}")
    browser_version = wd.capabilities.get("browserVersion")
    if not browser_version:
        # Chrome reports the browser version as just 'version'.
        browser_version = wd.capabilities.get("version")
    log(f"Browser Version: {browser_version}")
    log(f"Selenium API Version: {selenium.__version__}")
    if chrome_caps := wd.capabilities.get("chrome"):
        version_str = chrome_caps.get("chromedriverVersion")
        log(f"Chromedriver version: {version_str}")
    wd.set_script_timeout(120)
    log(f"Current URL: {wd.current_url}")
    wd.get("https://example.com")
    log("Done")
    wd.quit()


i = 0
while True:
    i += 1
    log(f"=== Iteration {i} ===")
    main()
```
</details>

<details>
<summary>Container information</summary>

```
docker run --rm -it --add-host="host.docker.internal:host-gateway" python:3.11-slim-bullseye bash
# Digest: sha256:3d260508893319ee271989fe16c0ddf13c710ec5ce1d0708cda7bc30da8389e7

# Inside the container
apt-get update && apt-get install nano  # used to paste contents into repro.py
pip install selenium==4.7.2
python repro.py
```

`pip freeze` details:

```
async-generator==1.10
attrs==22.2.0
certifi==2022.12.7
h11==0.14.0
idna==3.4
outcome==1.2.0
PySocks==1.7.1
selenium==4.7.2
sniffio==1.3.0
sortedcontainers==2.4.0
trio==0.22.0
trio-websocket==0.9.2
urllib3==1.26.14
wsproto==1.2.0
```

</details>

<details>
<summary>Relevant `docker-compose.yml` contents for Grid</summary>

```yaml
x-health: &healthcheck
  interval: 5s
  timeout: 5s
  retries: 3
  start_period: 10s

x-common: &common
  restart: unless-stopped
  pull_policy: always
  logging:
    options:
      max-size: "5M"
      max-file: "10"
  extra_hosts:
    host.docker.internal: host-gateway
  healthcheck:
    <<: *healthcheck

x-selenium-env: &selenium_env
  # Common env vars for Selenium nodes.
  SE_NODE_SESSION_TIMEOUT: 360
  SE_EVENT_BUS_HOST: selenium_hub
  SE_EVENT_BUS_PUBLISH_PORT: 4442
  SE_EVENT_BUS_SUBSCRIBE_PORT: 4443
  SCREEN_WIDTH: 1920
  SCREEN_HEIGHT: 1080
  SCREEN_DEPTH: 24
  TZ: Australia/Adelaide

services:
  selenium_hub:
    <<: *common
    image: selenium/hub:4
    container_name: selenium_hub
    ports:
      - "4444:4444"
    environment:
      SE_SESSION_REQUEST_TIMEOUT: 180
    healthcheck:
      <<: *healthcheck
      test: "wget -q -O - http://localhost:4444/status || exit 1"

  selenium_chrome: &selenium_chrome
    <<: *common
    image: selenium/node-chrome:4
    container_name: selenium_chrome
    volumes:
      - /dev/shm:/dev/shm
    environment:
      <<: *selenium_env
      SE_NODE_MAX_SESSIONS: "7"
    healthcheck:
      <<: *healthcheck
      test: "wget -q -O - http://localhost:5555/status || exit 1"

  selenium_firefox:
    <<: *selenium_chrome
    image: selenium/node-firefox:4
    container_name: selenium_firefox
    environment:
      <<: *selenium_env
      SE_NODE_MAX_SESSIONS: "1"
```
</details>

### Client Logs

- The first time I ran it, I received an unrelated `org.openqa.selenium.WebDriverException: Timed out waiting for driver server to shutdown.` error after about 95 iterations (I don't usually see this error, so I believe it was transient).
- I ran it again and it stalled on iteration 72.
- It does not appear that any other Selenium sessions were active in this instance, but it may be more prevalent if running the script multiple times in parallel.
- Times are in UTC+10:30, and I gave it about 9 hours before sending the keyboard interrupt.

<details>
<summary>Traceback in Python</summary>

```
2023-01-30 14:38:37.735963 === Iteration 72 ===
2023-01-30 14:38:43.810311 Browser Name: chrome
2023-01-30 14:38:43.810405 Browser Version: 109.0.5414.74
2023-01-30 14:38:43.810436 Selenium API Version: 4.7.2
2023-01-30 14:38:43.810462 Chromedriver version: 109.0.5414.74 (e7c5703604daa9cc128ccf5a5d3e993513758913-refs/branch-heads/5414@{#1172})
2023-01-30 14:38:43.837276 Current URL: data:,
^CTraceback (most recent call last):
  File "//repro.py", line 40, in <module>
    main()
  File "//repro.py", line 31, in main
    wd.get("https://example.com")
  File "/usr/local/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 455, in get
    self.execute(Command.GET, {"url": url})
  File "/usr/local/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 442, in execute
    response = self.command_executor.execute(driver_command, params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 294, in execute
    return self._request(command_info[0], url, body=data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 316, in _request
    response = self._conn.request(method, url, body=body, headers=headers)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/urllib3/request.py", line 78, in request
    return self.request_encode_body(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/urllib3/request.py", line 170, in request_encode_body
    return self.urlopen(method, url, **extra_kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/urllib3/poolmanager.py", line 376, in urlopen
    response = conn.urlopen(method, u.request_uri, **kw)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
                       ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py", line 449, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/local/lib/python3.11/site-packages/urllib3/connectionpool.py", line 444, in _make_request
    httplib_response = conn.getresponse()
                       ^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/http/client.py", line 1374, in getresponse
    response.begin()
  File "/usr/local/lib/python3.11/http/client.py", line 318, in begin
    version, status, reason = self._read_status()
                              ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/http/client.py", line 279, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/socket.py", line 706, in readinto
    return self._sock.recv_into(b)
           ^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```

</details>

**Comment by diemol** [comment]
Do you think you can help us by providing something we can use to reproduce? Probably the complete docker-compose file with the dependencies in a GitHub repository would suffice.

**Comment by ShadowLNC** [comment]
Just a quick update - we discovered further issues with the VMs, and
rebuilt them (moving from Ubuntu to Debian in the process).

I'm continuing to monitor the test runs and will retry the script from my
last message when I get a chance over the coming days, and will report back
with more details.

**Comment by ShadowLNC** [comment]
Hi - apologies for the delay.

It seems like our situation is mostly resolved, but the underlying issue may still be present.

### Timeline of events (all UTC+1030):

- 24 Jan
  - 0530: Cron script upgrades Selenium Grid (Hub+Node) instances to 4.8.0 (Chrome version 109)
- 25 Jan
  - 1800: First known test stall
- 27 Jan
  - 1700: Routine security patching to VMs (noting at the time that it may help with the Selenium issue)
- (28-29 Jan limited test runs due to it being a weekend)
- 30 Jan
  - Across the day: Identified elevated stall rates, filed this issue
- 31 Jan
  - Across the day: Identified elevated error rates in tests and slow test runs, some were Selenium tests, some not
- 01 Feb
  - 1030: Attempted to pin to Selenium Grid to 4.7, to no avail
  - 1100: VM behaviour consistent with resource starvation, yet all metrics show <70% memory and <70% CPU consumption. All test runs now stopped due to these issues.
  - 1530: Begin replacing VMs with Debian image (was Ubuntu) - using Selenium Grid 4.8
  - 2300: Restored most VMs and resuming test runs
- 02 Feb
  - 0900: Restored remaining VMs
  - Across the day: Selenium is stable, no other issues
- 09 Feb
  - Across the day: Upstream network issues (VMs on-prem)
  - 1238: First of 6 stalls occur
  - 1407: Last of 6 stalls occur (across multiple VMs)

### Current State

Since those stalls on 09 Feb, we have not seen any further issues with Selenium. The VMs were rebuilt with Docker Compose V2, but the compose file and tests (Selenium commands) are consistent with the [repro script attached in my earlier comment](https://github.com/SeleniumHQ/selenium/issues/11596#issuecomment-1409484271).

### Additional Tracebacks from 09 Feb

<details>
<summary>Stalled at 12:38:49, SIGINT sent at 13:32:27</summary>

```text
EXCEPTION: (KeyboardInterrupt)
Traceback (most recent call last):
 <...snip custom application code to call WebElement.find_elements...>
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 892, in find_elements
   return self.execute(Command.FIND_ELEMENTS, {"using": by, "value": value})["value"] or []
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 442, in execute
   response = self.command_executor.execute(driver_command, params)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 294, in execute
   return self._request(command_info[0], url, body=data)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 316, in _request
   response = self._conn.request(method, url, body=body, headers=headers)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 78, in request
   return self.request_encode_body(
          ^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 170, in request_encode_body
   return self.urlopen(method, url, **extra_kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/poolmanager.py", line 375, in urlopen
   response = conn.urlopen(method, u.request_uri, **kw)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 699, in urlopen
   httplib_response = self._make_request(
                      ^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 445, in _make_request
   six.raise_from(e, None)
 File "<string>", line 3, in raise_from
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 440, in _make_request
   httplib_response = conn.getresponse()
                      ^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 1374, in getresponse
   response.begin()
 File "/usr/local/lib/python3.11/http/client.py", line 318, in begin
   version, status, reason = self._read_status()
                             ^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 279, in _read_status
   line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/socket.py", line 706, in readinto
   return self._sock.recv_into(b)
          ^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```
</details>

<details>
<summary>Stalled at 12:39:12, SIGINT sent at 13:35:28</summary>

```text
EXCEPTION: (KeyboardInterrupt)
Traceback (most recent call last):
 <...snip custom application code to call selenium.webdriver.remote.webdriver.WebDriver.refresh()...>
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 684, in refresh
   self.execute(Command.REFRESH)
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 442, in execute
   response = self.command_executor.execute(driver_command, params)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 294, in execute
   return self._request(command_info[0], url, body=data)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 316, in _request
   response = self._conn.request(method, url, body=body, headers=headers)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 78, in request
   return self.request_encode_body(
          ^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 170, in request_encode_body
   return self.urlopen(method, url, **extra_kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/poolmanager.py", line 375, in urlopen
   response = conn.urlopen(method, u.request_uri, **kw)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 699, in urlopen
   httplib_response = self._make_request(
                      ^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 445, in _make_request
   six.raise_from(e, None)
 File "<string>", line 3, in raise_from
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 440, in _make_request
   httplib_response = conn.getresponse()
                      ^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 1374, in getresponse
   response.begin()
 File "/usr/local/lib/python3.11/http/client.py", line 318, in begin
   version, status, reason = self._read_status()
                             ^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 279, in _read_status
   line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/socket.py", line 706, in readinto
   return self._sock.recv_into(b)
          ^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```
</details>

<details>
<summary>Stalled at 12:40:03, SIGINT sent at 13:37:46</summary>

```text
EXCEPTION: (KeyboardInterrupt)
Traceback (most recent call last):
  <...snip custom application code to call selenium.webdriver.remote.webdriver.WebDriver.get()...>
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 455, in get
   self.execute(Command.GET, {"url": url})
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 442, in execute
   response = self.command_executor.execute(driver_command, params)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 294, in execute
   return self._request(command_info[0], url, body=data)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 316, in _request
   response = self._conn.request(method, url, body=body, headers=headers)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 78, in request
   return self.request_encode_body(
          ^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 170, in request_encode_body
   return self.urlopen(method, url, **extra_kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/poolmanager.py", line 375, in urlopen
   response = conn.urlopen(method, u.request_uri, **kw)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 699, in urlopen
   httplib_response = self._make_request(
                      ^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 445, in _make_request
   six.raise_from(e, None)
 File "<string>", line 3, in raise_from
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 440, in _make_request
   httplib_response = conn.getresponse()
                      ^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 1374, in getresponse
   response.begin()
 File "/usr/local/lib/python3.11/http/client.py", line 318, in begin
   version, status, reason = self._read_status()
                             ^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 279, in _read_status
   line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/socket.py", line 706, in readinto
   return self._sock.recv_into(b)
          ^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```
</details>

<details>
<summary>Stalled at 13:22:19, SIGINT sent at 16:34:50 (traceback appears to be the same as the 12:40:03 instance)</summary>

```text
EXCEPTION: (KeyboardInterrupt)
Traceback (most recent call last):
 <...snip custom application code to call selenium.webdriver.remote.webdriver.WebDriver.get()...>
   self.driver.get(url)
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 455, in get
   self.execute(Command.GET, {"url": url})
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 442, in execute
   response = self.command_executor.execute(driver_command, params)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 294, in execute
   return self._request(command_info[0], url, body=data)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 316, in _request
   response = self._conn.request(method, url, body=body, headers=headers)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 78, in request
   return self.request_encode_body(
          ^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 170, in request_encode_body
   return self.urlopen(method, url, **extra_kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/poolmanager.py", line 375, in urlopen
   response = conn.urlopen(method, u.request_uri, **kw)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 699, in urlopen
   httplib_response = self._make_request(
                      ^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 445, in _make_request
   six.raise_from(e, None)
 File "<string>", line 3, in raise_from
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 440, in _make_request
   httplib_response = conn.getresponse()
                      ^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 1374, in getresponse
   response.begin()
 File "/usr/local/lib/python3.11/http/client.py", line 318, in begin
   version, status, reason = self._read_status()
                             ^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 279, in _read_status
   line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/socket.py", line 706, in readinto
   return self._sock.recv_into(b)
          ^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```
</details>

<details>
<summary>Stalled at 14:07:53, SIGINT sent at 16:37:02 (traceback appears to be the same as the 12:40:03 instance)</summary>

```text
EXCEPTION: (KeyboardInterrupt)
Traceback (most recent call last):
 <...snip custom application code to call selenium.webdriver.remote.webdriver.WebDriver.get()...>
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 455, in get
   self.execute(Command.GET, {"url": url})
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 442, in execute
   response = self.command_executor.execute(driver_command, params)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 294, in execute
   return self._request(command_info[0], url, body=data)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 316, in _request
   response = self._conn.request(method, url, body=body, headers=headers)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 78, in request
   return self.request_encode_body(
          ^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 170, in request_encode_body
   return self.urlopen(method, url, **extra_kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/poolmanager.py", line 375, in urlopen
   response = conn.urlopen(method, u.request_uri, **kw)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 699, in urlopen
   httplib_response = self._make_request(
                      ^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 445, in _make_request
   six.raise_from(e, None)
 File "<string>", line 3, in raise_from
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 440, in _make_request
   httplib_response = conn.getresponse()
                      ^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 1374, in getresponse
   response.begin()
 File "/usr/local/lib/python3.11/http/client.py", line 318, in begin
   version, status, reason = self._read_status()
                             ^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 279, in _read_status
   line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/socket.py", line 706, in readinto
   return self._sock.recv_into(b)
          ^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```
</details>

<details>
<summary>Stalled at 13:03:05, SIGINT sent at 16:39:06</summary>

```text
EXCEPTION: (KeyboardInterrupt)
Traceback (most recent call last):
 <...snip custom application code to call selenium.webdriver.remote.webdriver.WebDriver.current_url...>
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 538, in current_url
   return self.execute(Command.GET_CURRENT_URL)["value"]
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/webdriver.py", line 442, in execute
   response = self.command_executor.execute(driver_command, params)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 294, in execute
   return self._request(command_info[0], url, body=data)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/selenium/webdriver/remote/remote_connection.py", line 316, in _request
   response = self._conn.request(method, url, body=body, headers=headers)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 74, in request
   return self.request_encode_url(
          ^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/request.py", line 96, in request_encode_url
   return self.urlopen(method, url, **extra_kw)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/poolmanager.py", line 375, in urlopen
   response = conn.urlopen(method, u.request_uri, **kw)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 699, in urlopen
   httplib_response = self._make_request(
                      ^^^^^^^^^^^^^^^^^^^
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 445, in _make_request
   six.raise_from(e, None)
 File "<string>", line 3, in raise_from
 File "/app/.venv/lib/python3.11/site-packages/urllib3/connectionpool.py", line 440, in _make_request
   httplib_response = conn.getresponse()
                      ^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 1374, in getresponse
   response.begin()
 File "/usr/local/lib/python3.11/http/client.py", line 318, in begin
   version, status, reason = self._read_status()
                             ^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/http/client.py", line 279, in _read_status
   line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/local/lib/python3.11/socket.py", line 706, in readinto
   return self._sock.recv_into(b)
          ^^^^^^^^^^^^^^^^^^^^^^^
KeyboardInterrupt
```
</details>

All tracebacks reported the following information (presumably from the Grid connection):
```text
Build info: version: '4.8.0', revision: '267030adea'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '4.19.0-23-amd64', java.version: '11.0.17'
```

### Theory and Possible Resolutions

I believe that an underlying network/Docker/Linux issue causes/caused the calls from Python to Selenium Hub not to complete. (I'm not sure whether the request packets were dropped, or the response packets.)

I'm guessing that our Ubuntu VMs were already on the precipice of instability and the Selenium Grid 4.8 upgrade may have increased resource usage slightly, just tipping over into an intermittent issue, which was greatly exacerbated by our security patching (to the point where Selenium Grid 4.7 was also affected).

It seems like the following line in the Python client bindings (line 54 in `/selenium/webdriver/remote/remote_connection.py` for version 4.7.2, [line 55 on the trunk](https://github.com/SeleniumHQ/selenium/blob/trunk/py/selenium/webdriver/remote/remote_connection.py#L55)) may be responsible:

```python3
        return None if cls._timeout == socket._GLOBAL_DEFAULT_TIMEOUT else cls._timeout
```

Assuming that `None` is being supplied upon socket creation or to [`socket.settimeout()`](https://docs.python.org/3/library/socket.html#socket.socket.settimeout) (instead of `socket._GLOBAL_DEFAULT_TIMEOUT`), then I believe that `socket.recv_into()` could block indefinitely if the connection is held open but the specific packets are lost due to some system issue as described above.

I think we may be able to apply a workaround in our codebase as follows:

```python3
import selenium.webdriver
selenium.webdriver.remote.remote_connection.RemoteConnection.set_timeout(30)
```

I tested and verified that `socket._GLOBAL_DEFAULT_TIMEOUT` does not change even when calling `socket.setdefaulttimeout()` - it seems this is a sentinel value and doesn't actually hold the default timeout. I wonder if it is possible to change `RemoteConnection.get_timeout()` to directly return `cls._timeout`, but I don't know if that value has to be serialised somewhere, which would prevent us from returning `socket._GLOBAL_DEFAULT_TIMEOUT`.

In any case, returning `socket._GLOBAL_DEFAULT_TIMEOUT` does not directly address the issue, as Python's default socket behaviour also appears to be an indefinite timeout; it only allows Selenium to respect changes from calling `socket.setdefaulttimeout()`. I don't know whether Selenium would consider using a different default to guard against this scenario described in this ticket, but this would continue the current behaviour where Selenium is prevented from respecting changes from calling `socket.setdefaulttimeout()`.

For now, we will continue to monitor the situation. If we see tests stalling again, we'll try the workaround I've listed. I will report back if we continue to see issues despite the workaround.

Please feel free to use this ticket to discuss/fix socket default timeout, and/or close as "won't fix" - thanks for your assistance. 😄

**Comment by diemol** [comment]
Thank you for the detailed report. However, it is not clear what needs to be done and it points more towards the infrastructure rather than Selenium?

What we can do is close this issue, and if you that the workaround helps after tests failed again, please open a pull request and reference this issue so we can determine how to add it to the code base.
